---
title: "Analysis of taxonomic data from DRiDO microbiome study"
output: rmarkdown::github_document
---

_Author:_ Lev Litichevskiy

_Date:_ October 30, 2023

In this tutorial, we import taxonomic classification results and perform several basic analyses:

1. ɑ-diversity: influence of age and dietary restriction (DR)
2. Uniqueness: influence of age and DR
3. PCoA plot
4. Examples of microbes affected by age and DR

N.B. This tutorial is derived from `kraken.Rmd`.

```{r setup, include=FALSE}
knitr::opts_chunk$set(dpi=300,fig.width=7)
```

# Load libraries

```{r, warning=F, message=F}
library(tidyverse)
library(phyloseq)
library(speedyseq) # faster tax_glom
library(ggpubr) # for t-test

diet.palette <- c(AL="seashell4",
                  `1D`="skyblue",
                  `2D`="royalblue4",
                  `20`="orange",
                  `40`="firebrick")
```

# Import Kraken data

This data has been aggregated by `stool.ID`, meaning that we aggregated samples originating from the same stool sample. We aggregated by summing Kraken counts.

```{r}
agg.kraken.df <- read.table(
  "../data/kraken_matrix_agg_by_stool_ID_n1303x2997.txt", 
  sep="\t", header=T, row.names=1)
dim(agg.kraken.df)
```

1303 taxons x 2997 stool IDs

```{r}
agg.kraken.df[1:5, 1:4]
```

# Import metadata

`stool.meta.df` contains information about each stool sample, `mouse.meta.df` contains information about the mice.

```{r}
stool.meta.df <- read.table(
  "../data/metadata/stool_metadata_after_QC_no_controls_n2997_230620.txt", 
  sep="\t", header=T)
mouse.meta.df <- read.csv("../data/metadata/AnimalData_Processed_20230712.csv")
stool.meta.annot.df <- stool.meta.df %>%
  merge(mouse.meta.df, by.x="mouse.ID", by.y="MouseID")

dim(stool.meta.annot.df)
n_distinct(stool.meta.annot.df$stool.ID)
```

2997 (unique) stool IDs

What kind of metadata do we have?

```{r}
stool.meta.annot.df %>% 
  dplyr::select(stool.ID, mouse.ID, sample.type, age.wks, DOE, SurvDays) %>% 
  head()
```

There was one stool collection at 5 months (prior to initiation of DR at 6 months, or 25 weeks). We'll create a metadata field that indicates that these 5-month samples were all from mice feeding _ad libitum_ (AL).

```{r}
stool.meta.annot.df <- stool.meta.annot.df %>% 
  mutate(Diet.5mo.as.AL=case_when(
    age.wks < 25 ~ "AL",
    TRUE ~ as.character(Diet)))
```

# Import Kraken taxonomy

```{r}
kraken.tax.df <- read.table(
  "../data/kraken_taxonomy_n1303.txt", 
  sep="\t", header=T, quote="", row.names=1)
dim(kraken.tax.df)
```

1303 taxons

```{r}
kraken.tax.df[1:5, ]
```

# Create phyloseq object

We will make our data into a `phyloseq` object because this package has several useful functions.

```{r}
kraken.sample.meta.df <- stool.meta.annot.df %>% column_to_rownames("stool.ID")

agg.physeq <- phyloseq(
  agg.kraken.df %>% as.matrix %>% otu_table(taxa_are_rows=T),
  kraken.sample.meta.df %>% sample_data,
  kraken.tax.df %>% as.matrix %>% tax_table
)
nsamples(agg.physeq)
```

2997 samples

# Aggregate to genera

```{r}
agg.physeq.genus <- agg.physeq %>% tax_glom(taxrank="genus")
```

```{r}
ntaxa(agg.physeq)
ntaxa(agg.physeq.genus)
```

1303 taxa --> 252 genera

```{r}
sum(otu_table(agg.physeq.genus)) / sum(otu_table(agg.physeq))
```

85% of reads could be assigned to the level of genus

# Compute distances

Bray-Curtis on genus-level relative abundances.

```{r}
agg.genus.bc.dist <- agg.physeq.genus %>%
  transform_sample_counts(function(x) x/sum(x)) %>%
  phyloseq::distance(method="bray")
```

# Alpha diversity

We'll look at two ɑ-diversity metrics: Shannon index and Simpson's index. Both of these metrics account for both richness and evenness.

```{r}
alpha.div.df.genus <- estimate_richness(
  agg.physeq.genus, 
  measures=c("Shannon", "Simpson")) %>% 
  rownames_to_column("stool.ID") %>% 
  merge(stool.meta.annot.df, by="stool.ID")

alpha.div.long.df.genus <- alpha.div.df.genus %>% 
  pivot_longer(c(Shannon, Simpson), names_to="metric")
```

## Versus age

```{r}
alpha.div.long.df.genus %>% 
  
  # omit the few 46 month samples
  dplyr::filter(age.approx.months <= 40) %>% 
  
  ggplot(aes(x=age.approx.months, y=value)) +
  
  # plot line of best fit
  geom_smooth(method="lm") +
  
  geom_boxplot(aes(group=age.approx.months), fill="seagreen") +
  
  # indicate that DR was initiated at 6 months
  geom_vline(xintercept=6, lty=2) +
  
  facet_wrap(~metric, scales="free") +
  labs(title="Alpha diversity versus age", x="Age (months)", y="") +
  theme_bw(base_size=10)
```

## Versus diet

```{r}
alpha.div.long.df.genus %>% 
  
  # omit the few 46 month samples for consistency with the versus age plot
  dplyr::filter(age.approx.months <= 40) %>% 
  
  # specify diet order
  mutate(Diet.5mo.as.AL=factor(Diet.5mo.as.AL, levels=c("AL", "1D", "2D", "20", "40"))) %>% 
  
  ggplot(aes(x=Diet.5mo.as.AL, y=value, fill=Diet.5mo.as.AL)) +
  geom_boxplot() +
  
  # t-test
  stat_compare_means(method="t.test", label="p.signif", ref.group="AL") +
  
  labs(x="", y="") +
  facet_wrap(~metric, scales="free") +
  scale_fill_manual(values=diet.palette) +
  theme_bw(base_size=10) +
  theme(legend.position="none")
```

# Uniqueness

Uniqueness is the distance from a microbiome sample to its nearest neighbor ([Wilmanski 2021](https://www.nature.com/articles/s42255-021-00348-0)).

```{r}
agg.genus.bc.dist.mat <- agg.genus.bc.dist %>% as.matrix
diag(agg.genus.bc.dist.mat) <- NA
agg.genus.uniq.bc.df <- data.frame(
  uniqueness=apply(agg.genus.bc.dist.mat, MARGIN=2, FUN=function(x) {min(x, na.rm=T)}),
  stool.ID=sample_names(agg.physeq.genus)) %>%
  merge(stool.meta.annot.df, by="stool.ID")
```

## Versus age

Regress uniqueness versus age to get slope and p-value.

```{r}
agg.genus.uniq.bc.lm.summary <- agg.genus.uniq.bc.df %>%
  dplyr::filter(age.approx.months <= 40) %>% # omit few 46 month samples
  lm(uniqueness ~ age.approx.months, data=.) %>% summary

agg.genus.uniq.bc.lm.pval <- agg.genus.uniq.bc.lm.summary$coefficients[
  "age.approx.months", "Pr(>|t|)"]
agg.genus.uniq.bc.lm.beta <- agg.genus.uniq.bc.lm.summary$coefficients[
  "age.approx.months", "Estimate"]
```

```{r}
agg.genus.uniq.bc.df %>% 
  
  # omit few 46 month samples
  dplyr::filter(age.approx.months <= 40) %>%
  
  ggplot(aes(x=age.approx.months, y=uniqueness)) +
  
  # indicate that DR was initiated at 6 months
  geom_vline(xintercept=6, lty=2) +
  
  # plot line of best fit
  geom_smooth(method="lm") +
  
  geom_boxplot(aes(group=age.approx.months), fill="seagreen") +
  
  labs(x="Age (months)", y="Uniqueness", 
       title=sprintf("Taxonomic uniqueness, slope = %.1e, p-value = %.1e",
                     agg.genus.uniq.bc.lm.beta, agg.genus.uniq.bc.lm.pval)) +
  theme_bw(base_size=10) +
  theme(legend.position="none")
```

## Versus diet

```{r}
agg.genus.uniq.bc.df %>% 
  
  # omit the few 46 month samples for consistency with the versus age plot
  dplyr::filter(age.approx.months <= 40) %>% 
  
  # specify diet order
  mutate(Diet.5mo.as.AL=factor(Diet.5mo.as.AL, levels=c("AL", "1D", "2D", "20", "40"))) %>%
  
  ggplot(aes(x=Diet.5mo.as.AL, y=uniqueness, fill=Diet.5mo.as.AL)) +
  geom_boxplot() +
  
  # t-test
  stat_compare_means(method="t.test", label="p.signif", ref.group="AL") +
  labs(x="Dietary group", y="Uniqueness", title="Taxonomic uniqueness") +
  scale_fill_manual(values=diet.palette) +
  theme_bw(base_size=10) +
  theme(legend.position="none")
```

# PCoA

We'll make a principal coordinates analysis (PCoA) plot to look for the overall influence of age and DR in this dataset.

```{r}
agg.pcoa.genus.bc <- agg.genus.bc.dist %>% 
  cmdscale(eig=T, k=3)

agg.pcoa.genus.bc.df <- merge(
  data.frame(agg.pcoa.genus.bc$points) %>% setNames(c("PCoA1", "PCoA2", "PCoA3")),
  kraken.sample.meta.df,
  by="row.names") %>% 
  mutate(Diet=factor(Diet, levels=c("AL", "1D", "2D", "20", "40")),
         Diet.5mo.as.AL=factor(Diet.5mo.as.AL, levels=c("AL", "1D", "2D", "20", "40")))
```

```{r}
frac.var.explained.by.pcos <- agg.pcoa.genus.bc$eig / sum(agg.pcoa.genus.bc$eig)
frac.var.explained.by.pcos[1:3]
```

PCoA1 and PCoA2 explain 35% and 8% of overall variance, respectively

```{r}
agg.pcoa.genus.bc.df %>% 
  
  # color by diet, size by age
  ggplot(aes(x=PCoA1, y=PCoA2, color=Diet.5mo.as.AL, size=age.approx.months)) +
  
  # slight transparency
  geom_point(alpha=0.8) +
  
  scale_color_manual(values=diet.palette) +
  
  # customize the legend for size
  scale_size_continuous(breaks=c(5,10,16,22,28,34,40,46), range=c(0.25,2)) +
  
  labs(color="Diet", size="Age (months)") +
  theme_bw(base_size=10)
```

We can sort of see a color gradient (red mostly at the top) and a size gradient (smaller dots at the bottom)

# Examples of individual genera

## Bifidobacterium versus age

Bifidobacterium is the genus that increases the most with age.

```{r}
agg.physeq.genus %>%
  
  # convert to relative abundances
  transform_sample_counts(function(x) x/sum(x)) %>% 
  
  # extract data from phyloseq object
  psmelt %>% 
  
  # take log10
  mutate(log10relab=log10(Abundance)) %>% 
  
  # subset to Bifido
  dplyr::filter(genus == "Bifidobacterium") %>% 
  
  # omit the few 46 month samples
  dplyr::filter(age.approx.months <= 40) %>%
  
  ggplot(aes(x=age.approx.months, y=log10relab)) +
  
  # indicate that DR was initiated at 6 months
  geom_vline(xintercept=6, lty=2) +
  
  geom_smooth(method="lm") +
  geom_boxplot(aes(group=age.approx.months), fill="seagreen") +
  labs(x="Age (months)", y="log10(relab)", title="Bifidobacterium") +
  theme_bw(base_size=10)
```

## Ligilactobacillus versus diet

Ligilactobacillus is strongly increased by DR

```{r}
agg.physeq.genus %>% 
  
  # convert to relative abundances
  transform_sample_counts(function(x) x/sum(x)) %>% 
  
  # extract data from phyloseq object
  psmelt %>% 
  
  # subset to Bifido
  dplyr::filter(genus == "Ligilactobacillus") %>% 
  
  # omit the few 46 month samples
  dplyr::filter(age.approx.months <= 40) %>%
  
  # specify diet order
  mutate(Diet.5mo.as.AL=factor(Diet.5mo.as.AL, levels=c("AL", "1D", "2D", "20", "40"))) %>% 
  
  # take log10
  mutate(log10relab=log10(Abundance)) %>% 
  
  ggplot(aes(x=Diet.5mo.as.AL, y=log10relab, fill=Diet.5mo.as.AL)) +
  geom_boxplot() +
  
  # t-test
  stat_compare_means(ref.group="AL", label="p.signif", method="t.test") +
  
  labs(x="", y="log10(relab)", title="Ligilactobacillus", fill="Diet") +
  scale_fill_manual(values=diet.palette) +
  theme_bw(base_size=10)
  theme(legend.position="none")
```

# sessionInfo

```{r}
sessionInfo()
```

